### PR TITLE
Interface implementations

### LINKED ISSUE #18: Implement Input template class
Implement Input template class that takes a type T and a c string. This class should construct implicit from T and should have a converting operator to T. The string should be parsed at compile time.

### LINKED ISSUE #25: Input- and OutputProxy names
I find the names InputConnector and OutputConnector more suiting for the current names InputProxy and OutputProxy. May I rename them?

### LINKED ISSUE #28: Move implementation into a namespace
namespace name should be yats

### PR DESCRIPTION
**Pull request by Highlife1911**
This PR contains the basic interface implementations.

Fixes #25 
Fixes #18
Fixes #28

### CODE REVIEW

File: source/yats/include/yats/Input.h [hunk 1]
```diff
@@ -1,7 +1,27 @@
+
 #pragma once
 
 
+/// <summary>
+/// <para>Main class to represent an input</para>
+/// <para><c>T</c> Type of input value</para>
+/// <para><c>Id</c> Unique identifier of input</para>
+/// </summary>
+template<typename T, size_t Id>
```
Comment (lawben) [inline source/yats/include/yats/Input.h]: We wanted to change this to `uint64_t`, didn't we?

File: source/yats/include/yats/Output.h [hunk 2]
```diff
@@ -1,16 +1,29 @@
...
 
+template<typename... Args>
+using OutputBundle = std::tuple<Args...>;
 
+/// <summary>
+/// <para>Main class to represent an output</para>
+/// <para><c>T</c> Type of output value</para>
+/// <para><c>Id</c> Unique identifier of output</para>
+/// </summary>
+template<typename T, size_t Id>
```
Comment (lawben) [inline source/yats/include/yats/Output.h]: `uint64_t`?

File: source/yats/include/yats/TaskConfigurator.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+
+/**/
+class AbstractTaskConfigurator
+{
+public:
+	AbstractTaskConfigurator() = default;
+	virtual ~AbstractTaskConfigurator() = default;
+
+	virtual std::unique_ptr<AbstractNodecontainer> make() const = 0;
```
Comment (lawben) [inline source/yats/include/yats/TaskConfigurator.h]: Abstract**Task**Container

File: source/yats/include/yats/TaskConfigurator.h [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+	TaskConfigurator()
+	{
+		parseInputParameters();
+		parseOutputParameters();
+	}
+
+	std::unique_ptr<AbstractNodecontainer> make() const override
+	{
+		return std::make_unique<Nodecontainer<Node>>();
```
Comment (lawben) [inline source/yats/include/yats/TaskConfigurator.h]: **Task**Container

File: source/yats/include/yats/TaskConfigurator.h [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+	virtual InputConnector& input(const std::string& name) = 0;
+	virtual InputConnector& input(size_t id) = 0;
+
+	virtual OutputConnector& output(const std::string& name) = 0;
+	virtual OutputConnector& output(size_t id) = 0;
+};
+
+
+template <typename Node>
```
Comment (lawben) [inline source/yats/include/yats/TaskConfigurator.h]: `template <typename Task>`

File: source/yats/include/yats/TaskConfigurator.h [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+class AbstractTaskConfigurator
+{
+public:
+	AbstractTaskConfigurator() = default;
+	virtual ~AbstractTaskConfigurator() = default;
+
+	virtual std::unique_ptr<AbstractNodecontainer> make() const = 0;
+
+	virtual InputConnector& input(const std::string& name) = 0;
+	virtual InputConnector& input(size_t id) = 0;
```
Comment (lawben) [inline source/yats/include/yats/TaskConfigurator.h]: This should be a `uint64_t` according to our discussion yesterday. Also, as we are exposing this to the user, it should be `typedef`ed (or `using ...`) to something like `InputID` so this actually has a semantic meaning.

File: source/yats/include/yats/TaskConfigurator.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		return m_outputs.at(id);
+	}
+
+protected:
+	void parseInputParameters()
+	{
+		parseInputParameter<0>();
+	}
+
+	template<size_t Index>
```
Comment (lawben) [inline source/yats/include/yats/TaskConfigurator.h]: `uint64_t` for all templates here

File: source/yats/include/yats/TaskConfigurator.h (hunk 6 above)
Comment (Highlife1911) [reply source/yats/include/yats/TaskConfigurator.h]: we might want to introduce a using like using Identifier = uint64_t

File: source/yats/include/yats/TaskConfigurator.h (hunk 7 above)
Comment (Highlife1911) [reply source/yats/include/yats/TaskConfigurator.h]: size_t is fine here.

**Comment by Highlife1911** [comment]
@lawben Can you remove the namespace yats from the unit tests?

**Comment by lawben** [comment]
@Highlife1911 removed it, but what exactly is the purpose? I think it is a lot less convenient to remove the namespace and prepend `yats::` to everything. Especially for our `_id` this is ugly, as we have to add `using namespace yats;` in each scope where this occurs.

**Comment by Highlife1911** [comment]
I would not put _id into the namespace.

**Comment by Highlife1911** [comment]
you have to add std::make_tuple to the return statement because creating a tuple from {} is C++17.

**Comment by lawben** [comment]
@Highlife1911 @Zanthelo @Shimrra @mflueggen 
As #33 covers this and more, I will close this PR due to massive conflicts. If there is something that was changed here, but not in #33, we should change that separately. Now, we have a more or less basic interface from which we can start developing in multiple directions. :clap:

**Pull request closed without merging**